app/models/concerns/completed_run.rb

```diff
@@ -74,5 +74,9 @@ def total_playtime_ms(timing)
     def completed?(timing)
       duration_ms(timing).present? && duration_ms(timing) > 0
     end
+
+    def completed_runs_count(timing)
+      histories.to_a.count { |history_element| !history_element.duration_ms(timing).zero? }
```

Comment:
This code fetches all history entries into memory then counts them. We should be able to have the database do the counting for us, which will be much faster and reduce the risk of running out of memory, with something like
```sql
case timing
when Run::REAL
  histories.where('realtime_duration_ms > 0').count
when Run::GAME
  histories.where('gametime_duration_ms > 0').count
```
This may be also be better laid out as a scope on RunHistory ([see examples on Run][1]), which will give us a more Rubyesque syntax from the caller like
```ruby
@run.histories.completed.count
```
and still provide the same benefit.

[1]: https://github.com/glacials/splits-io/blob/master/app/models/run.rb#L31-L36

Reply:
`scope :by_category, ->(category) { where(category: category) }`
Is the `(category)` part a variable that you can pass is?  Never worked with scopes before so I don't understand the syntax that well.

Reply:
Yeah, `scope` is a method that takes two arguments, a name for the scope and a lambda. Lambdas are a Ruby (and other languages) construct and are sort of like code blocks. The stuff in the parens is the argument list to the lambda, just like the stuff in `|pipes|` is the argument list to a code block.

It could be alternatively written as
```ruby
scope(:by_category, lambda { |category| where(category: category) })
```